Additional input refactoring

ceski-1 opened the pull request:
Reference: https://github.com/fabiangreffrath/woof/pull/1710#issuecomment-2132721825

ceski-1 commented:
@gendlin Please try this branch and let me know the results. This reverts https://github.com/fabiangreffrath/woof/commit/2fb776726aebf5693993359a8903c0f12d02b00c and https://github.com/fabiangreffrath/woof/commit/045b3a2797a14f3112b93d792a34c2721990dfd8, so the input code is identical to https://github.com/fabiangreffrath/woof/commit/46bc8de65843e8f08cce1ce97940b95ada41a215, which you described as working correctly.

gendlin commented:
This one runs fine for me.

ceski-1 commented:
@gendlin Try once more now that I've added back independent `raw_input` and `fake_longtics`.

gendlin commented:
Still fine.

ceski-1 commented:
Okay, thank you. I'm not sure what to conclude here. This branch is now at feature parity with the master branch, with the only difference being static function calls in this branch and function pointer calls in the master branch. While the issue seems resolved, I can't reproduce it, so it's possible that this could be inadvertently triggered again in the future and I won't know why. Any insights @fabiangreffrath @rfomin ?

gendlin commented:
If you have a version of the input refactoring with more granular atomic commits, I could bisect through that to further pinpoint the issue. But if no-one else can repro, maybe not worth the trouble,

gendlin commented:
Maybe if you just break out the mouse accel piece, since that seems like the most likely culprit. Mouse feels "muddy" around low speeds - might be a thresholding issue?

ceski-1 commented:
If you don't mind a bit of detective work, I'll put something together later with more granular commits. Thank you for your patience.

rfomin commented:
> This branch is now at feature parity with the master branch, with the only difference being static function calls in this branch and function pointer calls in the master branch.

I don't see where the problem is yet, but I don't mind reverting the refactoring. It might be better to remove other things from g_game.c, like functions related to demo playback/compatibility (I have such plans).

ceski-1 commented:
Okay here is what I should have done originally, which is break down the refactoring into smaller commits. @gendlin Please try the latest commit from this branch and bisect if needed. Thank you.

gendlin commented:
Ok, first bad commit is 8ee4837 Simplify input calculations

ceski-1 commented:
> Ok, first bad commit is [8ee4837](https://github.com/fabiangreffrath/woof/commit/8ee48377f96e86df395f0fe9030e470645005eb2) Simplify input calculations

I dropped that commit, and adjusted the ones after, starting from here: https://github.com/fabiangreffrath/woof/pull/1712/commits/e0c83ac3433af7f83de770b373a31b5fbd65b7ac

How is it now?

gendlin commented:
First bad commit is 38116c1 Reduce scope

fabiangreffrath commented:
> First bad commit is [38116c1](https://github.com/fabiangreffrath/woof/commit/38116c1bcc104880b21bd90e745c493e03171d09) Reduce scope

He, how is that possible?

gendlin commented on src/g_input.c:
```diff
@@ -46,7 +46,7 @@ void G_ClearLocalView(void)
 }
 
-void G_UpdateLocalView_FakeLongTics(void)
+static void UpdateLocalView_FakeLongTics(void)
```
Removing `static` on this line fixes the latency problem😐

gendlin replied:
`__declspec(noinline)` also works

ceski-1 replied:
How is the mouse input when not using fake_longtics? Let's say with uncapped and raw_input enabled. Does your bisecting lead back to this commit as well?

gendlin replied:
raw_input when not using fake_longtics has always been good, it doesn't seem to be affected by any of this.

gendlin commented:
Part of the flakiness here might be due to all the indirect calls through function pointers, which I think aren't very good nowadays. They're a barrier to inlining, and if you tickle the linker e.g. by changing the linkage in https://github.com/fabiangreffrath/woof/commit/38116c1bcc104880b21bd90e745c493e03171d09, it can layout your function in a less cache-friendly place in the binary in some totally unpredictable way. That's my best guess what is going on, anyways.

I think definitely think things like `G_UpdateLocalView` would be better replaced with a simple switch or if statement and/or a direct function call. The predictable branch is free and the direct call gets inlined most of the time - it's more performant and more deterministic than indirect calls these days, although not as cool :)

gendlin commented:
Part of the flakiness here might be due to all the indirect calls through function pointers, which I think aren't very good nowadays. They're a barrier to inlining and if you tickle the linker e.g. by changing the linkage in https://github.com/fabiangreffrath/woof/commit/38116c1bcc104880b21bd90e745c493e03171d09, it can layout your function in a less cache-friendly place in the binary in some totally unpredictable way. That's my best guess what is going on, anyways.

I think definitely think things like `G_UpdateLocalView` would be better replaced with a simple switch or if statement and/or a direct function call. The predictable branch is free and the call gets inlined most of the time if it's internal to the translation unit.

fabiangreffrath commented:
> Part of the flakiness here might be due to all the indirect calls through function pointers, which I think aren't very good nowadays. They're a barrier to inlining and if you tickle the linker e.g. by changing the linkage, it can layout your function in a less cache-friendly place in the binary in some totally unpredictable way. That's my best guess what is going on, anyways.

While this surely makes sense in some way, I doubt that the choice of inlining a single function or not should result in a human-perceivable latency. Especially, when you found out that *not* inlining that function gets rid of the latency. Compiler output is always unpredictable and depends on, well, the compiler. Which one do you use, btw?

> I think definitely think things like `G_UpdateLocalView` would be better replaced with a simple switch or if statement and/or a direct function call. The predictable branch is free and the call gets inlined most of the time if it's internal to the translation unit.

In general, branching is considered more expensive than dereferencing a function pointer. And even then, we are probably talking about sub-microseconds here, so I'm still not sure if this is really something that we could fix - or even manipulate in a predictable way - in code.

ceski-1 commented:
@gendlin I made new commits from https://github.com/fabiangreffrath/woof/pull/1712/commits/0261bd1587076afff0bf924b1899a21bac0c9651 to https://github.com/fabiangreffrath/woof/pull/1712/commits/8a8b8bdd5cf1b1588792278af5bef6bce1d1b83f. Please try it and let's see how it goes.

gendlin commented:
@ceski-1 All good. 8a8b8bd tests good.

ceski-1 commented:
For now, the input latency regression is resolved and there's some small additional refactoring, so I think this is ready for merging.

I'll stop investigating for now as it's beyond my current skill level to do anything more. I'll keep this PR in mind in case this issue reappears in the future. Thanks again for testing @gendlin.

fabiangreffrath commented:
> For now, the input latency regression is resolved and there's some small additional refactoring, so I think this is ready for merging.

Thank you! For those of us less involved into this topic, which change made the difference in the end?

> Thanks again for testing @gendlin.

Yes, thank you very much!

ceski-1 commented:
> For those of us less involved into this topic, which change made the difference in the end?

The main "fix" was keeping a few `localview` functions in g_game.c as shown here: https://github.com/fabiangreffrath/woof/pull/1712/commits/8b43225b84c6717d943386342cb00add18a87be3. I'm not entirely satisfied with that conclusion, but I'll live with it for now.

rfomin reviewed (approved):
I still can't reproduce the issue, but I'm not opposed to this changes.
